webapp/channels/src/components/advanced_text_editor/use_post_box_indicator.tsx

```diff
@@ -100,6 +100,7 @@ function useTimePostBoxIndicator(channelId: string) {
 
     const currentUserId = useSelector(getCurrentUserId);
     const isSelfDM = isDM && teammateId === currentUserId;
+    const isBot = Boolean(isDM && teammate?.is_bot);
```

Comment:
I wonder why this was not caught by the linter :thinking:

Reply:
yeah me neither, I removed the `?` from other places and ran a linter locally and there were no error. I'll lodge a ticket to investigate this further, seems like a simple null check that that should be caught by the linter. Something may have broken.

Reply:
Actually, I think the return type of `getUser` is overriding the specified type, checking this.

Reply:
Okay, so yeah, the type specified by called function overrides the type specified where the function is called, so in this case TS sees the variable's type as `UserProfile` and not `UserProfile | undefined`. Changing the type of `getUser` produces 22 errors across the codebase so I'm avoiding doing it right here. This PR will suffice for now.

Reply:
I've debated with myself the pros and cons of defining a type as object vs a Map and I think we can consider that as a solution as well, though I don't know all the aspects about it

Reply:
The problem is that in many places you will have to do stuff like:

```
function MyComponent(props: Props) {
    const user = useSelector(getCurrentUser);
    const myCallback = useCallback(() => {
        if (!user) {
            return;
        }
        ...
    }
    // And many more callbacks like that
    if (!user) {
        return;
    }
}
```

We could explore the possibility of creating special HOC, or wrapper, or however you want to call them, that would do something like this:

```
function ensureCurrentUser<T extends {currentUser: UserProfile}>(Component: ComponentType<T>) {
    return (props: Omit<'currentUser', T>) => {
        const currentUser = useSelector(getCurrentUser);
        if (!currentUser) {
            return null;
        }
        return <Component
            currentUser={currentUser}
            {...props}
        />
    }
}
```
Then it would be used like...
```
const MyComponent(props: Props) {...}
export default ensureCurrentUser(MyComponent);
```

We could potentially even add a fallback to the ensure functions, not to just not show it, but show some error message.

Reply:
I'd definitely like if we could more strictly null check these as well, but we definitely have to be conscious of how to make that not a headache to read or use as a developer. I like that we're forced to null check more thanks to TS and to the stricter type definitions that certain selectors have, but it unfortunately ends up making the code very noisy and full of `(all as any).sorts?.of && punctuatiation!` which makes reading and reasoning about it harder.

Approaches for reducing that include using HOCs like that, abstracting out Redux by passing important non-nullable things as props/context, or investigating the null object pattern more (which we already sort of do for Preferences and everywhere that a selector returns an empty array or map object when something is undefined). We likely can't use the same technique for everything since we have plenty of cases where we may want to load an object, show a loading indicator because we know someone else is loading it, or show a default value.

It's one of the things on my list that I do think we should look at, but because we have so many checks mitigating this in existing code already, I don't rank it as highly as some of the other tech debt things like reusable components or [making it easier to actually write tests without dozens of lines of Redux cruft](https://github.com/mattermost/mattermost/pull/29395/files#diff-661b3d0ce628cf1ad0acb6d372a4f429b582496d2a8ac51bc11492b366291ddfR12)